llmfoundry/optim/lion8b.py

```diff
@@ -0,0 +1,429 @@
+        if name in d:
+            if len(d) != 1:
+                raise ValueError(
+                    f'If state dict specifies {name}, it must not ' +
+                    f'specify other keys. Got {list(d.keys())}')
+            self.set_data(d[name])
+            return
+
+        self.quantized = d[f'{name}::quantized'].to(dtype=torch.int8)
+        self.scales = d[f'{name}::scales'].to(dtype=torch.float16)
+
+    def set_data(self, data: torch.Tensor) -> None:
+        if self._try_quantize:
+            if not data.is_cuda:
+                raise NotImplementedError(
+                    f'Attempting to quantize a non-CUDA {data.dtype} tensor ' +
+                    f'on device {data.device} with shape {data.shape}.')
+            self.data = None
+            assert self._f_encode is not None  # pyright
+            self.quantized, self.scales = self._f_encode(data)
+        else:
+            self.data = data.to(dtype=torch.float32)
+            self.quantized = None
+            self.scales = None
+
+    def is_quantized(self) -> bool:
+        return self.data is None
+
+    def materialize(self) -> torch.Tensor:
+        if not self.is_quantized():
+            assert self.data is not None  # pyright
+            return self.data
+        assert self._f_decode is not None  # pyright
+        assert self.quantized is not None  # pyright
+        assert self.scales is not None  # pyright
+        return self._f_decode(self.quantized, self.scales)
+
+    @property  # property to mirror Tensor interface
+    def is_cuda(self) -> bool:
+        if self.is_quantized():
+            assert self.quantized is not None  # pyright
+            return self.quantized.is_cuda
+        assert self.data is not None  # pyright
+        return self.data.is_cuda
+
+    @property  # property to mirror Tensor interface
+    def shape(self) -> Tuple[int]:
+        if self.is_quantized():
+            assert self.quantized is not None  # pyright
+            return self.quantized.shape
+        assert self.data is not None  # pyright
+        return self.data.shape
+
+    def numel(self) -> int:
+        if self.is_quantized():
+            assert self.quantized is not None  # pyright
+            return self.quantized.numel()
+        assert self.data is not None  # pyright
+        return self.data.numel()
+
+    def __repr__(self):
+        return (f'{self.__class__.__name__} quantized={self.is_quantized()} ' +
+                f'shape={self.shape}')
+
+
+def lion_step_unfused(grads: torch.Tensor,
+                      weights: torch.Tensor,
+                      momentums: torch.Tensor,
+                      lr: float,
+                      beta1: float,
+                      beta2: float,
+                      weight_decay: float = 0) -> torch.Tensor:
+    # f32 cast to match fused impl + for compatibility with f32 grads or weights
+    momentums = momentums.to(dtype=torch.float32)
```

Comment:
Just for my own comprehension - every parameter is upcast during the lerp and then downcast before moving onto the next parameter?

Reply:
Yes, basically. The parameters are kept in their original dtype, but we ensure the momentums are always f32 during the update (even if they were quantized). If they weren't quantized, this upcast is a no-op.